server/text_generation_server/utils/layers.py

```diff
@@ -690,13 +718,10 @@ def get_cos_sin(
             sin = torch.index_select(self._sin_cached, 0, position_ids)
             return cos.unsqueeze(1), sin.unsqueeze(1)
 
-        def forward(self, x: torch.Tensor, cos: torch.Tensor, sin: torch.Tensor):
-            rotary_dim = cos.shape[-1]
-            x1 = x[..., :rotary_dim]
-            x2 = x[..., rotary_dim : 2 * rotary_dim]
-
-            rotary_emb.apply_rotary(x1, x2, cos, sin, x1, x2, False)
-            return x
+    if IS_CUDA_SYSTEM:
```

Comment:
Can't the switch be done in a true function instead? 

```python
def forward(...):
    if IS_CUDA:
    ...
```
The imports can stay where they are to trigger the import error if that's why you segemented the code like this.

Reply:
While very ugly, I used this approach to avoid a controlflow in the forward itself, but the impact of it may be very minimal indeed.

Given that TGI already uses some dubious coding practices (`global`, definition of classes in if/else or try/except, :smile:), I thought it was fine to do it that way but I am happy to move the switch inside the forward.

Reply:
Dubious, really ?
Do you have better solutions for running into arbitrary environments ? 

Globals are not necessarily bad either.
Good coding practices are not a religion.

Reply:
Sounds good @Narsil, I was just explaining the motivation of the approach of dynamically defining the `forward` depending on whether torch+cu or torch+rocm is used.

Happy to do the change and put the controlflow rather in the forward itself.

Reply:
Thank you, I moved the controlflow in the forward in https://github.com/huggingface/text-generation-inference/pull/1243/commits/8617d4795a64873732e76b03d24fa683d091a72b. Yes, pragmatic is a more relevant description!